sdk/python/lib/pulumi/runtime/resource.py

```diff
@@ -907,6 +909,15 @@ async def do_register() -> None:
             nonlocal opts
             opts = opts if opts is not None else ResourceOptions()
 
+            # If we have a package reference, we need to wait for it to resolve.
+            packageRef = None
+            if package is not None:
+                packageRef = await package
+                # If we have a package reference we can clear some of the resource options
+                if packageRef is not None:
+                    opts.plugin_download_url = None
+                    opts.version = None
```

Comment:
In my head, a dynamically bridged provider was uniquely identified by a name&version pair, just like other providers. Is that wrong?

Likewise, unless `BaseProviderSpec` has a pluginDownloadURL option, we will need to keep that information so the base provider can set a plugin download URL.

Reply:
> In my head, a dynamically bridged provider was uniquely identified by a name&version pair, just like other providers. Is that wrong?

It is. We need every provider to have a unique name. But if a program sends a RegisterResourceRequest to the engine with just type:"myParamaterisedPackage:mod:typ" the engine can't find "myParamaterisedPackage" from just the name. It needs the parameter value as well.

> Likewise, unless BaseProviderSpec has a pluginDownloadURL option

It should do, I just haven't added it yet in this PR but there is a TODO for it.